panther_battery/src/adc_node.cpp

```diff
@@ -0,0 +1,191 @@
+  auto V_temp_sum = 0.0;
+  auto V_temp_bat_2 = 0.0;
+
+  try {
+    for (int i = 0; i < trials_num; i++) {
+      V_temp_sum += adc0_reader_->GetADCMeasurement("in_voltage0_raw", 0.0, 0.002);
+      std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    }
+
+    V_temp_bat_2 = V_temp_sum / trials_num;
+
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR(
+      this->get_logger(),
+      "Error reading ADC data: %s. "
+      "\nThe number of batteries cannot be determined. The single battery was adopted",
+      e.what());
+    return 1;
+  }
+
+  return V_temp_bat_2 > bat02_detect_thresh_ ? 1 : 2;
+}
+
+void ADCNode::IOStateCB(const IOStateMsg & msg) { charger_connected_ = msg.charger_connected; }
+
+void ADCNode::BatteryPubTimerCB()
+{
+  try {
+    V_bat_1_ = adc1_reader_->GetADCMeasurement("in_voltage0_raw", 0.0, 0.02504255);
+    V_bat_2_ = adc1_reader_->GetADCMeasurement("in_voltage3_raw", 0.0, 0.02504255);
+    I_bat_1_ = adc1_reader_->GetADCMeasurement("in_voltage2_raw", 625.0, 0.04);
+    I_bat_2_ = adc1_reader_->GetADCMeasurement("in_voltage1_raw", 625.0, 0.04);
+    V_temp_bat_1_ = adc0_reader_->GetADCMeasurement("in_voltage1_raw", 0.0, 0.002);
+    V_temp_bat_2_ = adc0_reader_->GetADCMeasurement("in_voltage0_raw", 0.0, 0.002);
+    I_charge_bat_1_ = adc0_reader_->GetADCMeasurement("in_voltage3_raw", 0.0, 0.005);
+    I_charge_bat_2_ = adc0_reader_->GetADCMeasurement("in_voltage2_raw", 0.0, 0.005);
+    last_battery_info_time_ = this->get_clock()->now();
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR(this->get_logger(), "Error reading ADC data: %s. ", e.what());
+  }
+
+  auto header_stamp = this->get_clock()->now();
+
+  if (
+    (this->get_clock()->now() - last_battery_info_time_) >
+    rclcpp::Duration::from_seconds(battery_timeout_)) {
+    battery_pub_->PublishUnknown(header_stamp);
+    if (battery_count_ == 2) {
+      battery_1_pub_->PublishUnknown(header_stamp);
+      battery_2_pub_->PublishUnknown(header_stamp);
+    }
+  } else {
+    auto temp_bat_1 = VoltageTempToDeg(V_temp_bat_1_);
+    auto temp_bat_2 = VoltageTempToDeg(V_temp_bat_2_);
+
+    if (battery_count_ == 2) {
+      battery_pub_->Publish(
+        header_stamp, (V_bat_1_ + V_bat_2_) / 2.0, (temp_bat_1 + temp_bat_2) / 2.0,
+        -(I_bat_1_ + I_bat_2_) + I_charge_bat_1_ + I_charge_bat_2_,
+        I_charge_bat_1_ + I_charge_bat_2_, charger_connected_);
+      battery_1_pub_->Publish(
+        header_stamp, V_bat_1_, temp_bat_1, -I_bat_1_ + I_charge_bat_1_, I_charge_bat_1_,
+        charger_connected_);
+      battery_2_pub_->Publish(
+        header_stamp, V_bat_2_, temp_bat_2, -I_bat_2_ + I_charge_bat_2_, I_charge_bat_2_,
+        charger_connected_);
+    } else {
+      battery_pub_->Publish(
+        header_stamp, V_bat_1_, temp_bat_1, -(I_bat_1_ + I_bat_2_) + I_charge_bat_1_,
+        I_charge_bat_1_, charger_connected_);
+    }
+  }
+
+  // print error messages
+  if (battery_pub_->HasErrorMsg()) {
```

Comment:
```suggestion
  if (battery_count_ == 1 && battery_pub_->HasErrorMsg()) {
```
Since all possible fault flags will come from the average of the two batteries the errors will be later logged separately on each battery. This will just repeat with shlight delay the error for the "abstract, mean battery"

Reply:
Done